File: DiningPhilosophers/EventManager.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace DiningPhilosophers
{
    static class EventManager
    {
        private static ConcurrentDictionary<string, EventArgs> _eventHandlers = new ConcurrentDictionary<string, EventArgs>();

        public static void Broadcast(string eventName, string arg = null)
        {
            _eventHandlers[eventName].Mutex.WaitOne();
            foreach (var callback in _eventHandlers[eventName].Callbacks)
            {
                new Task(() => callback(arg)).Start();
            }
            if (_eventHandlers[eventName].AutoUnsubscribe)
            {
                _eventHandlers[eventName].Callbacks = new List<Action<string>>();
            }
            _eventHandlers[eventName].Mutex.ReleaseMutex();
        }

        public static void Subscribe(string name, Action<string> callback, bool autoUnsubscribe = true)
        {
            if (_eventHandlers.ContainsKey(name))
            {
                _eventHandlers[name].Mutex.WaitOne();
                _eventHandlers[name].AutoUnsubscribe = autoUnsubscribe;
                _eventHandlers[name].Callbacks.Add(callback);
                _eventHandlers[name].Mutex.ReleaseMutex();
            }
            else
            {
                var args = new EventArgs
                {
                    AutoUnsubscribe = autoUnsubscribe
                };
                args.Callbacks.Add(callback);
                _eventHandlers[name] = args;
            }
        }
    }

    public class EventArgs
    {
        public List<Action<string>> Callbacks = new List<Action<string>>();

        public bool AutoUnsubscribe = true;

        public Mutex Mutex = new Mutex();
    }
}

File: go/eventmanager.go
package main

import (
	"sync"

	cmap "github.com/orcaman/concurrent-map"
)

type EventArgs struct {
	Callbacks       []*func(string)
	AutoUnsubscribe bool
	Mux             sync.Mutex
}

type EventManager struct {
	eventHandlers cmap.ConcurrentMap
}

var (
	eventManagerSingleton *EventManager
	once                  sync.Once
)

func GetEventMangager() *EventManager {
	once.Do(func() {
		eventManagerSingleton = &EventManager{eventHandlers: cmap.New()}
	})

	return eventManagerSingleton
}

func (manager *EventManager) Broadcast(eventName string, arg string) {
	value, _ := manager.eventHandlers.Get(eventName)
	eventArgs, _ := value.(*EventArgs)
	eventArgs.Mux.Lock()
	defer eventArgs.Mux.Unlock()
	for _, callback := range eventArgs.Callbacks {
		cFunc := *callback
		go cFunc(arg)
	}
	if eventArgs.AutoUnsubscribe {
		eventArgs.Callbacks = nil
	}

}

func (manager *EventManager) Subscribe(name string, callback *func(string), autoUnsubscribe bool) {
	value, contains := manager.eventHandlers.Get(name)
	if contains {
		args, _ := value.(*EventArgs)
		args.Mux.Lock()
		defer args.Mux.Unlock()
		args.AutoUnsubscribe = autoUnsubscribe
		args.Callbacks = append(args.Callbacks, callback)

	} else {
		args := new(EventArgs)
		args.AutoUnsubscribe = autoUnsubscribe
		args.Callbacks = []*func(string){callback}
		manager.eventHandlers.Set(name, args)
	}
}
